agent/provider/src/dir.rs

```diff
@@ -0,0 +1,117 @@
+        bail!("Not a provider data directory: {}", dir.as_ref().display());
+    }
+    Ok(clean_dir(dir, 2, lifetime, dry_run))
+}
+
+fn is_provider_dir<P: AsRef<Path>>(dir: P) -> Result<bool> {
+    let mut files = vec![
+        (HARDWARE_JSON, false),
+        (PRESETS_JSON, false),
+        (GLOBALS_JSON, false),
+    ];
+
+    dir.as_ref()
+        .read_dir()?
+        .filter_map(|r| r.ok())
+        .map(|e| e.file_name().to_string_lossy().to_string())
+        .for_each(|name| {
+            for (n, v) in files.iter_mut() {
+                if name == *n {
+                    *v = true;
+                }
+            }
+        });
+
+    Ok(files.iter().all(|pair| pair.1))
+}
+
+fn clean_dir<P: AsRef<Path>>(dir: P, min_depth: usize, lifetime: Duration, dry_run: bool) -> u64 {
+    let mut dirs = Vec::new();
+    let deadline = SystemTime::now() - lifetime;
+
+    let total_bytes = WalkDir::new(dir.as_ref())
+        .min_depth(min_depth)
+        .into_iter()
+        .filter_map(|result| result.ok())
+        .filter_map(|entry| match entry.metadata() {
+            Ok(meta) => Some((entry.path().to_owned(), meta)),
+            _ => None,
+        })
+        .inspect(|path_meta| {
+            if path_meta.1.is_dir() {
+                dirs.push(path_meta.0.clone());
+            }
+        })
+        .filter(|path_meta| {
+            path_meta.1.is_file()
+                && match path_meta.1.modified() {
+                    Ok(sys_time) => sys_time <= deadline,
+                    _ => false,
+                }
+        })
+        .fold(0, |acc, path_meta| {
+            if !dry_run {
+                if let Err(_) = std::fs::remove_file(&path_meta.0) {
+                    return acc;
+                }
+            }
+            acc + path_meta.1.len()
+        });
+
+    if !dry_run {
+        dirs.sort_by_key(|path| path.components().count());
+        dirs.into_iter().rev().for_each(|path| {
+            if let Ok(mut contents) = path.read_dir() {
+                if contents.next().is_none() {
+                    let _ = std::fs::remove_dir_all(path);
+                }
+            }
+        });
+    }
+
+    total_bytes
+}
+
+fn parse_expr<S: AsRef<str>>(expr: S) -> Result<Duration> {
+    let expr = expr.as_ref();
+    let cap: Captures = match (*EXPR_REGEX).captures_iter(expr.trim()).next() {
+        Some(cap) => match cap.len() {
+            3 => cap,
+            _ => bail!("Invalid expression: {}", expr),
+        },
+        None => bail!("Unrecognized expression: {}", expr),
+    };
+
+    let base: u64 = cap[1].parse().context("invalid number")?;
+    let mul: u64 = match &cap[2] {
+        "s" => 1,
+        "m" => 60,
+        "h" => 3600,
+        "d" | "" => 3600 * 24,
+        "w" => 3600 * 24 * 7,
+        "M" => 3600 * 24 * 30,
+        "y" => 3600 * 24 * 365,
+        u => bail!("Invalid unit: {}", u),
+    };
+
+    Ok(Duration::from_secs(base * mul))
+}
```

Comment:
Humantime https://docs.rs/humantime/2.0.1/humantime/ is what you need here ;)
It is already dependency of provider